prevent-broker-scale-down-if-contains-partition .md

```diff
@@ -0,0 +1,60 @@
+
+## Current situation
+
+Currently, when the number of Kafka replicas is decreased in the Kafka CR the StatefulSet will shut down the highest numbered brokers and there is no check to see if these brokers contain any topic-partition replicas.
+The [documentation](https://strimzi.io/docs/operators/latest/configuring.html#scaling-clusters-str) recommends that all topic-partition replicas are reassigned before doing this scaling action, which means that if someone tries to scale down without re-assigning the brokers, it can cause data loss.
+
+## Motivation
+
+We should definitely introduce some sort of logic which can detect if the broker which is going to be removed still contains the partitions or not.
+If the partitions are still available on the broker which is going to be scaled, then we should get some warning in the logs which will prompt users to do the reassignment and hopefully reduce the chance of data loss.
+
+## Proposal
+
+This proposal suggest how we can add the check to detect if the broker still contains any partitions and what to do if the broker scale down is not possible.
+
+Checking if the broker contains partition
+The best way to detect if the broker contains any partitions is to query the topics.
+To do this, we will have to make use of the `AdminClient` class.
+We should probably put the logic for this check inside the `scaleDown` method present in the `KafkaReconciler` class.
+The `scaleDown` method gets triggered when we scale down the no. of replicas in the `Kafka CR`.
+So this check will target the broker which are going to be removed ie. the pods with the highest number.
+
+Methods to be created in the process:
+
+- `getTopicNameAndDescription` (Boolean Type) - This method will be placed inside the `scaleDown()` method in `KafkaReconciler` class and will provide us with the Collection of `TopicDescription`.
+- `describeTopics()` and `topicNames()`  - These two methods will provide us with the description and names of all the topics ie first we will get the names of the topics using the `topicNames` method and using these names we will extract the description of the topics.
+- `brokerHasAnyTopics()` - This method will take as input a podId and a `Collection<TopicDescription>`.
+  It will go through all the broker and see if any partitions are assigned to it.
+
+### Flow :
+
+When the replicas are changed inside the Kafka resource, reconciliation will happen.
+During this reconciliation, we move to the `scaleDown()` method present in the `KafkaReconciler` class.
+The `scaleDown()` method contains the complete mechanism on how the broker is scaled down. So there, we will first create our `adminClient` and provide it with the Secrets.
+Then we will call the `checkPartitions()` method which will give us the `Collection<TopicDescription>`. 
```

Comment:
You should elaborate on how the error handling will look like here when it fails to connect to the Kafka cluster.

Reply:
@scholzj Hi, do you mean what if any `RuntimeException` pops up and then how will I handle the process to not stop the reconciliation process?